### PR TITLE
kv: limit transaction size, len, and entry length, reduce retry times

### PR DESCRIPTION
**Pull request by coocood**
Enforce those limits to prevent TiKV run out of resource.

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @disksing PTAL

File: kv/kv.go [hunk 1]
```diff
@@ -67,12 +67,16 @@ type RetrieverMutator interface {
 }
 
 // MemBuffer is an in-memory kv collection, can be used to buffer write operations.
-type MemBuffer RetrieverMutator
+type MemBuffer interface {
+	RetrieverMutator
+	Size() int
```
Comment (shenli) [inline kv/kv.go]: Why's the different between Size and Len?

File: kv/memdb_buffer.go [hunk 2]
```diff
@@ -25,6 +25,13 @@ import (
 	"github.com/pingcap/tidb/terror"
 )
 
+// Those limits is enforced to make sure the transaction can be well handled by TiKV.
+var (
+	EntrySizeLimit  = 8 * 1024 * 1024
```
Comment (shenli) [inline kv/memdb_buffer.go]: In this PR: https://github.com/pingcap/tidb/pull/2425/files,  requestMaxSize is 4 * 1024 * 1024.

File: kv/memdb_buffer.go [hunk 3]
```diff
@@ -58,6 +65,7 @@ func (m *memDbBuffer) SeekReverse(k Key) (Iterator, error) {
 	} else {
 		i = &memDbIter{iter: m.db.NewIterator(&util.Range{Limit: []byte(k)}), reverse: true}
 	}
+	m.db.Len()
```
Comment (shenli) [inline kv/memdb_buffer.go]: Why call Len() and ignore return value?

File: server/server_test.go [hunk 4]
```diff
@@ -507,20 +506,6 @@ func runTestStatusAPI(c *C) {
 	c.Assert(data.GitHash, Equals, printer.TiDBGitHash)
 }
 
-func runTestMultiPacket(c *C) {
```
Comment (shenli) [inline server/server_test.go]: Why remove this?

File: kv/memdb_buffer.go (hunk 3 above)
Comment (coocood) [inline kv/memdb_buffer.go]: This line should not be added.

File: session.go [hunk 5]
```diff
@@ -223,10 +222,18 @@ func (s *session) doCommit() error {
...
+	if s.txn != nil && s.txn.Valid() {
+		txnSize = s.txn.Size()
+	}
 	err := s.doCommit()
 	if err != nil {
 		if s.isRetryableError(err) {
-			err = s.Retry()
+			// Transactions will retry 2 ~ 10 times.
+			// We make larger transactions retry less times to prevent cluster resource outage.
+			txnSizeRate := float64(txnSize) / float64(kv.BufferSizeLimit)
```
Comment (shenli) [inline session.go]: Should we set txnSizeRate = min(1, txnSizeRate) to make maxRetryCount > = 1?

File: kv/error.go [hunk 6]
```diff
@@ -57,6 +58,8 @@ var (
 	ErrInvalidTxn = terror.ClassKV.New(codeInvalidTxn, "invalid transaction")
 	// ErrTxnTooLarge is the error when transaction is too large, lock time reached the maximum value.
 	ErrTxnTooLarge = terror.ClassKV.New(codeTxnTooLarge, "transaction is too large")
+	// ErrEntryTooLarge is the error when a key value entry is too large.
+	ErrEntryTooLarge = terror.ClassKV.New(codeEntryTooLarge, "entry is too large")
```
Comment (shenli) [inline kv/error.go]: There is no usage for this error in this PR?

File: kv/memdb_buffer.go (hunk 2 above)
Comment (shenli) [inline kv/memdb_buffer.go]: Add comment for the three consts.

Comment (coocood) [inline kv/memdb_buffer.go]: For simplicity, this is not set as session variables, so I set it as the same as real TiKV, which is 8MB.

File: kv/kv.go (hunk 1 above)
Comment (coocood) [inline kv/kv.go]: I should move the comment written in `memDbBuffer` here.

File: kv/mem_buffer_test.go [hunk 7]
```diff
@@ -42,14 +42,12 @@ type testKVSuite struct {
 }
 
 func (s *testKVSuite) SetUpSuite(c *C) {
-	s.bs = make([]MemBuffer, 2)
-	s.bs[0] = NewRBTreeBuffer()
```
Comment (tiancaiamao) [inline kv/mem_buffer_test.go]: why RBTree removed?

File: session.go [hunk 8]
```diff
@@ -327,7 +334,7 @@ func (s *session) Retry() error {
 			return errors.Trace(err)
 		}
 		retryCnt++
-		if (s.maxRetryCnt != unlimitedRetryCnt) && (retryCnt >= s.maxRetryCnt) {
+		if !s.unlimitedRetryCount && (retryCnt >= maxCnt) {
 			return errors.Trace(err)
 		}
 		kv.BackOff(retryCnt)
```
Comment (shenli) [inline session.go]: Should we make huge txn backoff longer time?

File: kv/mem_buffer_test.go [hunk 9]
```diff
@@ -42,14 +42,12 @@ type testKVSuite struct {
...
-	s.bs[1] = NewMemDbBuffer()
+	s.bs = make([]MemBuffer, 1)
```
Comment (tiancaiamao) [inline kv/mem_buffer_test.go]: if just one element, why need it be a slice?

File: session.go (hunk 5 above)
Comment (tiancaiamao) [inline session.go]: Will txnSize > kv.BufferSizeLimit ?

File: session.go (hunk 8 above)
Comment (coocood) [inline session.go]: Usually, large transaction takes longer time to execute, adding backoff time doesn't make much difference.

File: kv/mem_buffer_test.go (hunk 7 above)
Comment (coocood) [inline kv/mem_buffer_test.go]: It is not used in our code, and it doesn't have Size and Len function.

File: kv/mem_buffer_test.go (hunk 9 above)
Comment (coocood) [inline kv/mem_buffer_test.go]: To keep the line of changes as minimum as possible.
Make be we need to add another MemBuffer implementation later.

File: session.go (hunk 5 above)
Comment (shenli) [inline session.go]: I am not sure. But if txnSize > kv.BufferSizeLimit, maxRetryCount will be negtive and we will meet an infinite loop.

File: kv/error.go (hunk 6 above)
Comment (coocood) [inline kv/error.go]: There is usage at memdb_buffer.go

File: server/server_test.go (hunk 4 above)
Comment (coocood) [inline server/server_test.go]: This test fails and doesn't make much sense if we enforce the entry and transaction limit.

File: session.go (hunk 5 above)
Comment (coocood) [inline session.go]: @shenli 
Then it will not do any retry.
`-1` doesn't mean unlimited here anymore.

File: session.go (hunk 8 above)
Comment (shenli) [inline session.go]: If a large txn failed, the system may need more time to recover. So should we wait a longer time? @disksing What's your opinion？

**Comment by tiancaiamao** [comment]
LGTM @disksing

**Comment by disksing** [comment]
LGTM

**Pull request merged**